Do not inline methods that never return

mikedn opened the pull request:
Methods that do not contain return blocks can't ever exit normally, they either throw or loop indefinitely. Inlining is not beneficial in such cases as it increases the code size without providing any speed benefits. In the particular case of throws the inlined code can easily be 10 times larger than the call site.

The call to `fgMoreThanOneReturnBlock` has been replaced with code that does the same thing but also detects the existence of at least one return block. This avoids walking the basic block list twice.

Note that `BBJ_RETURN` blocks are also generated for `CEE_JMP`. Methods exiting via `CEE_JMP` instead of `CEE_RET` will continue to be inlined (assuming they were inlined before this change).

omariom commented:
@mikedn 
Will it be like `throw` in terms of optimizations around of the callsites to such methods?

mikedn commented:
> Will it be like throw in terms of optimizations around of the callsites to such methods?

Yes. I'll post an example later.

Anyway, note the WIP title :smile:. This very likely needs more work. I'm just trying to see if it works and show something to the JIT team to get some feedback.

omariom commented:
Here is an especially bad example of inlining such methods:
It is from `FindEntry` of `Dictionary<TKey,TValue>` when `TKey` is a ref type.
```C#
private int FindEntry(TKey key) {
    if( key == null) {
        ThrowHelper.ThrowArgumentNullException(ExceptionArgument.key);
    }
...
```

```ASM
       test     rdi, rdi
       jne      G_M14733_IG03
       mov      rcx, 0xD1FFAB1E
       call     CORINFO_HELP_NEWSFAST
       mov      rdi, rax
       mov      ecx, 5
       call     System.ThrowHelper:GetArgumentName(int):ref
       mov      rbx, rax
       mov      ecx, 0x1C4ED
       mov      rdx, 0xD1FFAB1E
       call     CORINFO_HELP_STRCNS
       mov      rcx, rax
       call     System.Environment:GetResourceStringLocal(ref):ref
       mov      rbp, rax
       mov      rcx, rdi
       call     System.Exception:Init():this
       lea      rcx, bword ptr [rdi+32]
       mov      rdx, rbp
       call     CORINFO_HELP_ASSIGN_REF
       mov      dword ptr [rdi+132], 0xD1FFAB1E
       lea      rcx, bword ptr [rdi+144]
       mov      rdx, rbx
       call     CORINFO_HELP_ASSIGN_REF
       mov      dword ptr [rdi+132], 0xD1FFAB1E
       mov      dword ptr [rdi+132], 0xD1FFAB1E
       mov      rcx, rdi
       call     CORINFO_HELP_THROW

G_M14733_IG03:
```

mikedn commented:
Yeah, with this change that `FindEntry` code becomes:
```
G_M14720_IG02:
       test     rdi, rdi
       jne      SHORT G_M14720_IG03
       mov      ecx, 5
       call     System.ThrowHelper:ThrowArgumentNullException(int)
G_M14720_IG03:
```
It's still not quite right, I was expecting the conditional jump to be inverted and the call be moved to a cold block. Seems like `fgRemoveRestOfBlock` doesn't always do what I need.

mikedn commented:
Here's an example that works as expected:
```C#
static void ThrowArgumentNullException() {
    throw new ArgumentNullException();
}
static int Main(string[] args) {
    if (args == null) {
        ThrowArgumentNullException();
        Console.WriteLine("bad");
    }
    return args.Length;
}
```
Without this change the generated code is:
```
G_M45812_IG01:
       56                   push     rsi
       4883EC20             sub      rsp, 32
G_M45812_IG02:
       4885C9               test     rcx, rcx
       7522                 jne      SHORT G_M45812_IG03
       48B94806C64CFE7F0000 mov      rcx, 0x7FFE4CC60648
       E857EC415F           call     CORINFO_HELP_NEWSFAST
       488BF0               mov      rsi, rax
       488BCE               mov      rcx, rsi
       E87422DD5D           call     System.ArgumentNullException:.ctor():this
       488BCE               mov      rcx, rsi
       E8A4AAE95E           call     CORINFO_HELP_THROW
G_M45812_IG03:
       8B4108               mov      eax, dword ptr [rcx+8]
G_M45812_IG04:
       4883C420             add      rsp, 32
       5E                   pop      rsi
       C3                   ret
```
With this change the generated code is:
```
G_M45811_IG01:
       4883EC28             sub      rsp, 40
G_M45811_IG02:
       4885C9               test     rcx, rcx
       7408                 je       SHORT G_M45811_IG05
G_M45811_IG03:
       8B4108               mov      eax, dword ptr [rcx+8]
G_M45811_IG04:
       4883C428             add      rsp, 40
       C3                   ret
G_M45811_IG05:
       E8E2FBFFFF           call     Program:ThrowArgumentNullException()
       CC                   int3
```
If `fgRemoveRestOfBlock` is not set the generated code is:
```
G_M45811_IG01:
       56                   push     rsi
       4883EC20             sub      rsp, 32
       488BF1               mov      rsi, rcx
G_M45811_IG02:
       4885F6               test     rsi, rsi
       7517                 jne      SHORT G_M45811_IG03
       E8E6FBFFFF           call     Program:ThrowArgumentNullException()
       48B9683057D239020000 mov      rcx, 0x239D2573068
       488B09               mov      rcx, gword ptr [rcx]
       E86465DD5D           call     System.Console:WriteLine(ref)
G_M45811_IG03:
       8B4608               mov      eax, dword ptr [rsi+8]
G_M45811_IG04:
       4883C420             add      rsp, 32
       5E                   pop      rsi
       C3                   ret
```
The throw method is not inlined but there are various shortcomings:
* the throw call is not moved to a cold block
* `ecx` is killed by the call so it is moved to a call preserved register and that requires that register to be preserved
* dead code that follows the call isn't removed

mikedn commented:
Yay, only one test failed, that will be fun.

Diff summary:
```
Total bytes of diff: -109152
    diff is an improvement.
Top file improvements by size (bytes):
     -101463 : System.Private.CoreLib.dasm
       -3531 : System.Net.Http.dasm
       -1771 : System.Collections.Immutable.dasm
       -1653 : System.Runtime.Extensions.dasm
        -348 : System.Xml.ReaderWriter.dasm
7 total files with size differences.
Top method regessions by size (bytes):
          26 : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[__Canon,EventRegistrationTokenList][System.__Canon,System.Runtime.InteropServices.WindowsRuntime.WindowsRuntimeMarshal+EventRegistrationTokenList]:get_Item(ref):struct:this
          20 : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[Guid,Guid][System.Guid,System.Guid]:get_Item(struct):struct:this
          20 : System.Private.CoreLib.dasm - System.Collections.Generic.Dictionary`2[__Canon,Guid][System.__Canon,System.Guid]:get_Item(ref):struct:this
          16 : System.Private.CoreLib.dasm - System.Runtime.CompilerServices.JitHelpers:UnsafeEnumCast(ref):int
           1 : System.Collections.Immutable.dasm - System.Collections.Immutable.SortedInt32KeyNode`1[HashBucket][System.Collections.Immutable.ImmutableDictionary`2+HashBucket[System.__Canon,System.Int32]]:SetItem(int,struct,ref,byref,byref):ref:this
Top method improvements by size (bytes):
       -3461 : System.Private.CoreLib.dasm - System.ThrowHelper:IfNullAndNullsAreIllegalThenThrow(ref,int)
       -1617 : System.Private.CoreLib.dasm - System.Diagnostics.Tracing.PropertyValue:GetFactory(ref):ref
        -780 : System.Private.CoreLib.dasm - System.Text.StringBuilder:AppendFormatHelper(ref,ref,struct):ref:this
        -308 : System.Private.CoreLib.dasm - System.Threading.Tasks.GenericDelegateCache`2[__Canon,__Canon][System.__Canon,System.__Canon]:.cctor()
        -307 : System.Net.Http.dasm - System.Net.Http.WinHttpHandler:SetRequestHandleOptions(ref):this
1037 total methods with size differences.
```
Regressions are caused by `fgRemoveRestOfBlock` not always working like it happens in the case of `FindEntry`. Most of the improvements are in System.Private.CoreLib due to its high usage of throw helpers.

It may be worth noting that this works only if the throw helper wasn't marked as no-inline. If it's no-inline the JIT won't figure out that it does not return.

omariom commented:
@mikedn 
Another case to check. [`ThrowHelper` has methods that don't throw themselves](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/ThrowHelper.cs#L45) but call other methods that throw. Will the change work for such case?

mikedn commented:
> Another case to check. ThrowHelper has methods that don't throw themselves but call other methods that throw. List's indexer calls such helper. Will the change work for this case?

Sort of. The parameterless helper is inlined and the other one is not. Neither should be inlined but it may be problematic to treat the parameterless helper as no-return as well.

I'll see if something can be done about this once I figure out what's going on with `FindEntry`.

mikedn commented:
The `FindEntry` issue (and many similar ones) was caused by a lack of no-return information on the second attempt to inline `ThrowArgumentNullException`. The first time the `CORINFO_FLG_BAD_INLINEE` is set on the inline and the second time the JIT doesn't attempt to inline anymore and thus doesn't know that the inline is no-return. There needs to be a `CORINFO_FLG_NO_RETURN` flag to deal with this.

I'll update the diff summary, now there are no size regressions.

BruceForstall commented:
cc @AndyAyersMS

AndyAyersMS commented on src/jit/morph.cpp:
```diff
@@ -6953,6 +6958,10 @@ GenTreePtr          Compiler::fgMorphCall(GenTreeCall* call)
+        else if ((call->gtFlags & GTF_CALL_NO_RETURN) != 0)
+        {
+            szFailReason = "No return";
+        }
```
Can you explain what motivated this change? Is there some downside to tail calling a no-return method (eg larger calling sequence?)

mikedn replied:
The tail call code is part of the epilog but with my change no epilog is generated anymore because the BB is converted to `BBJ_THROW`. The net result is that a test failed because an exception was no longer being thrown.

AndyAyersMS replied:
This is exactly the kind of thing that I'd like to see in a comment. A pointer to the test case that failed w/o this exclusion would be useful here too.

AndyAyersMS commented on src/jit/morph.cpp:
```diff
@@ -7654,6 +7663,11 @@ GenTreePtr          Compiler::fgMorphCall(GenTreeCall* call)
+        (call->TypeGet() == TYP_VOID))
+    {
+        fgRemoveRestOfBlock = true;
+    }
```
Somebody else on the jit team can chime in, but there should be some way to indicate that the return value is not actually going to be produced, so that this restriction can be lifted.

mikedn replied:
This was primarily caused by an assert I hit in a test. It would seem that the call node needs to have `GTF_EXCEPT` if `fgRemoveRestOfBlock` is set. I find this confusing, how come `GTF_EXCEPT` is not always set on calls to begin with? Anyway, I didn't look more into this because such cases should be rare.

AndyAyersMS commented on src/jit/flowgraph.cpp:
```diff
@@ -5746,14 +5746,36 @@ void          Compiler::fgFindBasicBlocks()
+            impInlineInfo->inlineResult->NoteFatal(InlineObservation::CALLEE_DOES_NOT_RETURN);
+            return;
+        }
+
```
I worry that this might be too general. I've seen cases where people have written performance-sensitive code that is effectively `while (true) ...` and then relied on exceptions to terminate processing, and such methods can benefit from being inlined.

Ideally we'd like to check that the inlinee does very little other than throw an exception.

jkotas replied:
> I've seen cases where people have written performance-sensitive code that is effectively  while (true) ...  and then relied on exceptions to terminate processing, and such methods can benefit from being inlined

Exception throwing has always been super slow path. I do not think there is any benefit in inlining infinite loops that are terminated by throwing exceptions.

mikedn replied:
As @jkotas already pointed out exceptions are very slow anyway. The typical throw sequence already needs at least 3 calls - new, ctor and throw - and the throw itself is very expensive.

Note that this does not affect methods like this one:
```C#
void ThrowIfNull(object value) {
    if (value == null)
        throw new ArgumentNullException();
}
```
Not inlining this would be bad since you'd pay the cost of a call even if `value` is not null.
```
```

mikedn replied:
> and the throw itself is very expensive

Some numbers: if in my test I pass a `null` to the `Test` method then the execution time is 50 minutes instead of 500ms! That's 6000 times slower than the no-exception case.

AndyAyersMS replied:
Yes, exceptions are slow. But the point is that `X is postdominated by a throw' does not imply `X is not performance sensitive`.

This is a tricky thing to get right because there are cases where a relatively large amount of computation is done solely for the purpose of throwing a more detailed exception.

My preference would be to start off conservatively here, only matching simple methods with this heuristic, for instance methods that are loop-free and relatively small. This should catch most of the idiomatic throw helper cases.

mikedn replied:
> But the point is that X is postdominated by a throw does not imply X is not performance sensitive.

I had similar doubts myself back when I did this experiment but I just can't find a reasonable use case that could be impacted by this. That X needs to be small enough for the method to be an inline candidate and it has to contain a loop otherwise the performance X will be dwarfed by the throw. And then the loop performance needs to be better when inlined due to optimizations such as constant propagation, otherwise the only impact we could possible measure would be the delay of the loop start by the time taken by the call.

I'd love to see an example that shows that this is indeed a problem. The current implementation already fails to reject inlining in some cases where there's no reason to inline, I'm somewhat reluctant to impose additional limitations.

> for instance methods that are loop-free

I'll have to check but I'm not sure there's any loop information available during inlining. Probably the next best thing would be to just match a single BB of type BBJ_THROW, possibly preceded by a list of BBJ_ALWAYS BBs. To make it even more conservative we could match only BBJ_THROW containing a single statement.

> and relatively small

Wouldn't this be redundant as the inliner already takes code size into account?

AndyAyersMS replied:
Yes, the current inliner -- via LegacyPolicy -- is conservative with respect to size of inlinees. But I'd rather not bake that assumption in here. Going forward, policies will likely be somewhat less conservative. 

It would be simple enough here to just impose a limit on the number of BB's. Or turn this into an informational observation and let the policy decide whether or not to inline, and set the limits there.

AndyAyersMS commented on src/jit/inline.cpp:
```diff
@@ -697,7 +697,9 @@ void InlineResult::Report()
-        if ((m_Callee != nullptr) && (obs != InlineObservation::CALLEE_IS_NOINLINE))
+        if ((m_Callee != nullptr) && 
+            (obs != InlineObservation::CALLEE_IS_NOINLINE) && 
+            (obs != InlineObservation::CALLEE_DOES_NOT_RETURN))
```
I presume you're not propagating here so that the jit rereads the IL for each call? It would be good noet that here and link to the issue you opened for recording noinline deductions with the runtime.

As with my note above, we'd really like to ensure that the inlinee is small here to avoid repeatedly scanning a large inlinee that must throw.

mikedn replied:
> It would be good noet that here and link to the issue you opened for recording noinline deductions with the runtime.

You mean the issue you opened and I hijacked? :) - #6095

AndyAyersMS replied:
Yes.

AndyAyersMS commented:
@mikedn Thanks for this, it looks very promising.

Can you work up some micro-benchmarks for these cases? You can use the pattern from the inline microbenchmarks I checked in recently -- they are very sensitive to small changes in codegen.

Also a general perf run would be nice. This is something that should be available soon from automation but in the meantime I can do some runs for you or show you how to run it -- however I won't get to it until later in the day.

mikedn commented:
> Can you work up some micro-benchmarks for these cases?

I'll try. I doubt that we'll see anything significant in terms of performance, this is first and foremost a size optimization.

mikedn commented:
> I doubt that we'll see anything significant in terms of performance, this is first and foremost a size optimization.

Well, they are right when they say that micro-benchmarks can prove anything. This little [test](https://gist.github.com/mikedn/1249488f12b2cf7b3aa24a2bfabcae04) runs in 500ms without my fix and 375ms with my fix.

I'll see if I can make a JIT benchmark out of it. Hmm, project.json... shudder...

AndyAyersMS commented:
Please don't create a new project.json. Re-use one of the common ones we have set up for the JIT, like [this one](https://github.com/dotnet/coreclr/tree/master/tests/src/JIT/config/benchmark/project.json).

And [here's a template](https://github.com/dotnet/coreclr/blob/master/tests/src/JIT/Performance/CodeQuality/Inlining/ConstantArgsByte.cs) for what the test should look like....

mikedn commented:
> Please don't create a new project.json. Re-use one of the common ones we have set up for the JIT, like this one.

I know, I know, it's just that there are certain aspects of project.json and the build system that are annoying. Like the warnings generated by those test project.json...

I added a benchmark but I haven't been able to run it yet.

omariom commented:
@mikedn 
> And if someone has the bad idea of applying [MethodImpl(MethodImplOptions.NoInlining)] to ThrowArgumentNullException..

:point_down:
```
System.Convert::ThrowByteOverflowException 
System.Convert::ThrowCharOverflowException 
System.Convert::ThrowInt16OverflowException 
System.Convert::ThrowInt32OverflowException 
System.Convert::ThrowInt64OverflowException 
System.Convert::ThrowSByteOverflowException 
System.Convert::ThrowUInt16OverflowException 
System.Convert::ThrowUInt32OverflowException 
System.Convert::ThrowUInt64OverflowException 
System.Reflection.Assembly::LoadFrom 

System.IO.Compression.DeflateStream::ThrowCannotReadFromDeflateStreamException 
System.IO.Compression.DeflateStream::ThrowCannotWriteToDeflateStreamException 
System.IO.Compression.DeflateStream::ThrowStreamClosedException 
System.IO.Compression.GZipStream::ThrowStreamClosedException 
```

AndyAyersMS commented:
Thanks for the benchmark. 

Since `FindEntry` was one of the motivating cases, can we measure any performance improvement there?

mikedn commented:
> Since FindEntry was one of the motivating cases, can we measure any performance improvement there?

I tried a few things but the performance of FindEntry doesn't seem to be affected in any way by this change. I also played with a list indexer and that one is slightly faster with this change - 100,000,000 `get_Item` take 80ms without this change and 70ms with this change.

However, I don't think that the performance improvement shown by the previous benchmark and `get_Item` is directly related to this change. It seems that the JIT has an unrelated problem - sometimes throw blocks aren't moved to the cold region and this change happens to avoid that problem. If the throw block is already in the cold region then not inlining has no measurable performance impact..

omariom commented:
> take 80ms without this change and 70ms with this change.

With your change it is 12.5% faster 👍

mikedn commented:
> With your change it is 12.5% faster 

12.5% makes it sound like it is a lot. But it really isn't, at 3.10GHz that my CPU has the difference is less than half a cycle per `get_Item` call. I'm not even sure how to explain such a difference. The number of instructions is identical in both cases. The branch is always taken so it shouldn't be a branch prediction issue. Most likely the throw code being in the fall-through path has a small impact on instruction fetching & decoding.

AndyAyersMS commented:
Smaller code is generally preferable even if it's not faster. So I'm happy with the CS benefits you see from this change, and, barring surprises elsewhere, think it is a nice improvement. 

Here's my take on the work remaining to actually enable it:

* The LegacyPolicy is intended to embody Desktop 4.6.1 / CoreCLR 1.0 behavior, and we don't want it to act on this new information. So this change needs to be policy specific. That means:
  * The observation needs to be informational and the policy needs to decide if the observation causes inlining to fail.
  * LegacyPolicy needs to ignore this new observation (which I believe it will do by default)
  * We'll need a new policy that is based on LegacyPolicy that acts on this information. 
  * Propose this new policy as the default policy for post 1.0, and update the InlinePolicy factory method to make the new policy the default and allow fallback (via a `COMPlus` setting) to the LegacyPolicy.
* I still think we want some kind of size limit, say a basic block limit. If the new policy is based on the LegacyPolicy you'll get this for free. When I look at handling this observation for new policies I can decide what limit might be appropriate.
* Comments about why we don't propagate noinline back to the runtime for this case.
* Comments about why we can't handle methods with return values. 
* Extend the test case with an example that will break if the return value restriction is lifted.
* Extend the test case with an example showing how earlier noinline deductions can prevent this from kicking in. For instance if the inlinee is must throw and contains a STARG, we won't deduce must throw.
* Verification that LegacyPolicy codegen is unaffected by this change (eg no jit diffs)
* Some assessment of throughput (TP) impact. Expectation is that this will make the jit faster, but I'd like to be sure. We really should update the jitutils to support this kind of measurement (cc @russellhadley)
* Some assessment of overall code quality impact (CQ). We can use the CoreCLR perf tests for this. These are not as easy to run as they should be, unfortunately.

I'd be happy to help out, especially with the TP/CQ assessments, since our tooling there is not yet what it should be.

mikedn commented:
Thanks for the detailed work plan! For a start I'll tweak the matching code to see how it affects the jit diff results so we can get an idea about what needs to go in the new policy.

mikedn commented:
@AndyAyersMS I've made a number of changes:
* New `NoReturnLegacyPolicy` that extends the `LegacyPolicy` and handles the (now informational) does not return observation. The block count is included and currently inlining is rejected only if the count is 1. It turns out that there's no difference compared to the previous approach that ignored the block count.
* The "no return" call flag is now set directly by `fgFindBasicBlocks`, disregarding inliner's decision. The inliner can thus completely ignore the "no return" observation and fail inlining for a different reason. In that case we still want to set the "no return" flag as without it the caller's code quality may be worse.
* No longer reject tail calls. The main reason to set `fgRemoveRestOfBlock` is to avoid unnecessary register preservation across "no return" calls but this does not apply to tail calls. It may still be useful to reject tail calls because the tail call sequence is slightly larger than a normal call but for now let's minimize the number of unnecessary changes.
* Set `fgRemoveRestOfBlock` for non-void calls. This requires setting `GTF_EXCEPT` on the call node but I'm not sure it's enough, this needs more work.
* Rely on `PropagateNeverToRuntime` to stop propagation of "no return" inline failure to the runtime instead of adding another special casing to `Report()`.
* Added more comments.

Stuff yet to do:
* A COMPLUS option to force the use of the legacy policy
* Additional benchmarks/tests. For example it would be good to look at cases such as `System.Convert::ThrowInt32OverflowException`, AFAIR they were marked as "NoInline" because they were blocking other convert methods from being inlined. But no "NoInline" blocks propagation of "no return" information which isn't good.
* Whatever else I forgot/missed.

AndyAyersMS commented on src/jit/flowgraph.cpp:
```diff
@@ -5746,14 +5746,47 @@ void          Compiler::fgFindBasicBlocks()
+
+            impInlineInfo->iciCall->gtCallMoreFlags |= GTF_CALL_M_DOES_NOT_RETURN;
+
+            compInlineResult->NoteInt(InlineObservation::CALLEE_DOES_NOT_RETURN, static_cast<int>(fgBBcount));
```
I'd really like this to be a `NoteBool` so that this observation is policy neutral. That creates a complication on the policy end, since any rejection will now be the result of two observations, and (at least from the policy's viewpoint) there's no guarantee on observation order. What I'd suggest is that for the earlier observation, we add a DEBUG mode  "we have seen this observation" flag and assert that it's true when the later observation is made.

mikedn replied:
You mean that I should use the existing `NUMBER_OF_BASIC_BLOCKS` observation to communicate `fgBBCount`, right?

mikedn replied:
> we add a DEBUG mode "we have seen this observation" flag 

I wonder if we couldn't simply rely on m_BasicBlockCount being 0 since 0 isn't a valid value.

AndyAyersMS replied:
We currently don't fire off `NUMBER_OF_BASIC_BLOCKS` until after `fgFindBasicBlocks' returns.
Also LegacyPolicy does not remember the values of most observations, it tends to just act on things immediately.

So likely in the new policy you'll want to do two things: remember the value of the noreturn observation, and (in DEBUG) remember that the noreturn observation was made. Then when you get the basic block count observation, assert for the latter, then check the block count and the noreturn value and trigger the inlining failure.

mikedn replied:
Yep, I've come to the same conclusion. It seems a bit strange to react to the basic block count observation instead of the no return observation but it works for now. Maybe later we could report the bb count earlier, it seems to be reported unnecessarily late.

AndyAyersMS commented on src/jit/inline.def:
```diff
@@ -61,6 +61,7 @@ INLINE_OBSERVATION(STFLD_NEEDS_HELPER,        bool,   "stfld needs helper",
 INLINE_OBSERVATION(THROW_WITH_INVALID_STACK,  bool,   "throw with invalid stack",      FATAL,       CALLEE)
 INLINE_OBSERVATION(TOO_MANY_ARGUMENTS,        bool,   "too many arguments",            FATAL,       CALLEE)
 INLINE_OBSERVATION(TOO_MANY_LOCALS,           bool,   "too many locals",               FATAL,       CALLEE)
+INLINE_OBSERVATION(DOES_NOT_RETURN,           bool,   "does not return",               FATAL,       CALLEE)
```
This needs to be informational now since it may not lead to failures in all policies. Also, I've been keeping these in alphabetical order by category.

AndyAyersMS commented on src/jit/inlinepolicy.h:
```diff
@@ -155,6 +155,25 @@ class LegacyPolicy : public LegalPolicy
+// NoThrowInlineLegacyPolicy extends the legacy policy by rejecting
+// inlining of methods that never return because they throw.
+
+class NoReturnLegacyPolicy : public LegacyPolicy
```
Let's call this the `EnhancedLegacyPolicy` or something similar since there are other changes we might put in here down the road (eg handling STARG, pinned locals, etc...)

AndyAyersMS commented:
This is coming along nicely... left you a few more comments.

@dotnet/jit-contrib, any guidance you can offer here on the `GTF_EXECPT` flag? It seems like it should be set on most any call, definitely on calls to inlineable user methods, but that doesn't seem to be the case today. Is there some implicit understanding that calls can cause exceptions?

mikedn commented:
Rebased and addressed feedback (including `IsLegacyPolicy`). CI failures appear to be caused by the CI itself.

Do you happen to know more about `CALLSITE_CONDITIONAL_THROW`? I was looking at #2833 to make a benchmark out of it and now I'm wondering if this change won't uncover some JIT limitation. If a throw helper is no longer inlined then there will be no `CEE_THROW` to generate a `CALLSITE_CONDITIONAL_THROW` observation. But the call's basic block will be converted to `BBJ_THROW` so it will affect liveness pretty much the same way a normal throw does.

AndyAyersMS commented:
Updates look good, thanks. For now we should only change the default policy for CoreCLR, but leave it at is for now and I'll start some desktop testing for you.

I don't know much about `CALLSITE_CONDITIONAL_THROW`, so I'll have to go do some digging. At first blush it looks like it might only impact the legacy backend.

Looks like some kind of  infrastructure issue took out your latest test run, so let me restart the tests.

@dotnet-bot retest this please

AndyAyersMS commented:
Hmm, CI did not retest. Let me try again.

@dotnet-bot retest this please

roterdam commented:
Why does putting no inlining on ThrowArgumentNullException slow down the code? According to the theory you present it should be faster since there is less code to execute.

mikedn commented:
> For now we should only change the default policy for CoreCLR, but leave it at is for now and I'll start some desktop testing for you

@AndyAyersMS  Should the "no return" call flag be set only for CoreCLR as well?

AndyAyersMS commented:
Good question. Let me think about that one...

mikedn commented:
> Why does putting no inlining on ThrowArgumentNullException slow down the code? According to the theory you present it should be faster since there is less code to execute.

@roterdam If `NoInlining` is used the JIT doesn't attempt to analyze `ThrowArgumentNullException` and as a result the "no return" information won't be available to the caller. Without this information the caller has to preserve live registers across the call and you usually end up executing more code in the non-exception path. A few posts above I posted a piece of disassembly labeled "If fgRemoveRestOfBlock is not set the generated code is". In that you can see that there are a few additional instructions - `push esi`, `mov esi, ecx` and `pop esi`.

There are 3 different ways to deal with such throw helpers:
1. Inline the helper. This is bad for code size and sometimes it also makes the code slightly slower because that throw code, even if not run, can impact instruction caching/decoding/prefetching. This is what the JIT tends to do today.
2. Do not inline the helper but don't bother with the "no return" information. This does improve code size but the lack of "no return" information can affect performance due to additional instructions on the non-exception path. `NoInlining` will always trigger this case.
3. Do not inline the helper and propagate the "no return" information to the caller. This is the ideal situation as it keeps the code size down without impacting caller's non-exception path performance like 2 does. This is what the JIT will do after this change.

benaadams commented:
@mikedn I have a number of exception functions that may or may not trigger this path (if NoInlining was removed - changed set) https://github.com/aspnet/KestrelHttpServer/pull/962/commits/6b161f51d49ab87180fda4c16b3c8a10da6761bb

Item in particular would be
```csharp
public void RejectRequest(RequestRejectionReasons reason)
{
    var ex = BadHttpRequestException.GetException(reason); // large switch statement
    SetBadRequestState(ex);
    throw ex;
}
```
This should trigger item 3. with this change?

mikedn commented:
> This should trigger item 3. with this change?

It should but it's possible to end up in case 2 if the JIT decides to reject inlining before the "no return" information is computed. The most likely reason for an early rejection is probably `CALLEE_TOO_MUCH_IL` but I don't think that this will be a problem in your case, that method seems to be pretty small. 

IMO `NoInlining` should not be used in that code no matter what the JIT does. There are some relatively large methods there and saving a few bytes won't help. 

Saving code is useful in generics such as collections and LINQ because then those "few bytes" start to add up due to unshared generic instantiations. It may also be useful in small non-generic methods that themselves are like to be inlined and then drag the throw code with them.

benaadams commented:
> IMO NoInlining should not be used in that code no matter what the JIT does. There are some relatively large methods there and saving a few bytes won't help.

They do get executed several million times a second so more about reducing instruction cache size; though as you've highlighted in this PR its not as simple as that :)

mikedn commented:
> They do get executed several million times a second so more about reducing instruction cache size

Normally that shouldn't matter, if the JIT inlines them then the block should end up in the cold region. That said, it appears that sometimes this doesn't happen and the throw blocks are left in the middle of the method. Perhaps I should investigate that as well. Though after this change this won't matter too much anymore.

mikedn commented:
@AndyAyersMS Somewhat related to non-void return throw helpers - `CALLEE_LACKS_RETURN` is no longer useful after this change (except of course, in the legacy case). 

Even before this change the JIT was rejecting "no return" method if they had non-void return. I think this is because there's no mechanism to remove the rest of the tree (or even the entire expression tree). `fgRemoveRestOfBlock` removes only subsequent statements.

mikedn commented:
> I don't know much about CALLSITE_CONDITIONAL_THROW, so I'll have to go do some digging. At first blush it looks like it might only impact the legacy backend.

Presumably you're saying this because the legacy codegen contains a function named `checkLiveness`? It also seems to be related to qmarks since those are handled differently by the legacy backend.

I tried to remove that observation and the non-legacy JIT seems to be working fine. No asserts fired and diffs shows that a couple of additional methods were inlined and the generated code appears to be correct.

What seems even more strange is that this observation relies on `seenConditionalJump` being set previously. What happens if there's a conditional jump in the method but it is seen after `CEE_THROW` is imported?

AndyAyersMS commented:
Agreed, detection of the supposedly unsafe throw case is a bit dodgy. More reason to believe it's probably not an essential constraint.

AndyAyersMS commented:
@mikedn I have finally got through a round of desktop testing and perf testing. Things are looking good:

- Throughput
 - destkop jit throughputon par with current jit
-  Code Quality
 - desktop jit CQ (which measures time) is on par overall with current jit. There are some swings which seem likely to be loop top alignment issues (simpsn_d)
 - my local CoreCLR perf runs (which measure instructions retired) show no real change
 - some ASP.Net plaintext runs I did showed a small win, maybe 2%
- Code Size
 - your jit-diff measurement above show improvements
 - desktop SPMI runs agree and show good size improvements
- Correctness
 - no issues uncovered in desktop testing

@pgavlin has confirmed that GTF_EXCEPT is implicitly set on calls.

We should work on finishing this up. The biggest open issue is whether we need to flip the default and preserve pure legacy behavior by excluding the no return flag when not building for the CoreCLR. For now I'd recommend enabling clean fallback to legacy behavior but leaving the new behavior as default. This means the setting of the no return flag on the call becomes conditional on the inline policy. To avoid a specific coupling I'd expose this via a new method on the InlineStrategy -- when the strategy is constructed it can check which policy will be created and set a local flag bit, then `fgFindBasicBlocks` can call the new method to check whether the no return should be set on the call.

mikedn commented:
> pgavlin has confirmed that `GTF_EXCEPT` is implicitly set on calls.

Does that mean that `GTF_CALL` implies `GTF_EXCEPT`? Anyway, I still need to set `GTF_EXCEPT` on the call node otherwise I get an assert. I don't really like the non-void return stuff, I'm tempted to remove that part. I don't know any use cases where this could be helpful and doing it right requires additional work anyway (since it's not enough to remove statements, the rest of the tree should also be removed).

> To avoid a specific coupling I'd expose this via a new method on the InlineStrategy -- when the strategy is constructed it can check which policy will be created and set a local flag bit, then fgFindBasicBlocks can call the new method to check whether the no return should be set on the call.

Ok, I'll take a look tomorrow evening. Heading to bed now :)

pgavlin commented:
> pgavlin has confirmed that GTF_EXCEPT is implicitly set on calls.

I'm not sure that I confirmed this--I am pretty sure that `GTF_EXCEPT` still needs to be set explicitly. I may have stated that the value of `GTF_EXCEPT` for a call node is directly derivable from the node itself (unlike e.g. `GT_IND` nodes, in which the value of the exception flag depends on the address being dereferenced).

benaadams commented:
> I don't really like the non-void return stuff, I'm tempted to remove that part. I don't know any use cases where this could be helpful

Suppose it would be?
```csharp
Task DoSomething(bool success)
{
    if (success)
    {
        return Task.CompletedTask;
    }

    return ThrowError();
}

Task ThrowError()
{
    throw new Exception();
}
```
vs
```csharp
Task DoSomething(bool success)
{
    if (success)
    {
        return Task.CompletedTask;
    }

    ThrowError();
    // Can never reach here but compiler doesn't know
    return Task.CompletedTask; 
}

void ThrowError()
{
    throw new Exception();
}
```

mikedn commented:
@benaadams I think it's actually versus
```C#
Task DoSomething(bool success)
{
    if (!success)
        ThrowError();

    return Task.CompletedTask;
}

void ThrowError()
{
    throw new Exception();
}
```

mikedn commented:
> To avoid a specific coupling I'd expose this via a new method on the InlineStrategy -- when the strategy is constructed it can check which policy will be created and set a local flag bit, then fgFindBasicBlocks can call the new method to check whether the no return should be set on the call.

@AndyAyersMS Wouldn't be possible to simply use the existing `UsesLegacyPolicy` method? Or, if you think that it should be up to the policy to decide whether to propagate the no return information shouldn't the new method be added to the policy? It's not clear to me why the strategy would be involved here.

AndyAyersMS commented:
Yes, that would be fine -- the code runs in the context of a particular inline so can just query the current InlineResult.

mikedn commented:
Updated:
* Made `JitInlinePolicyLegacy` default to 1 for desktop builds so they continue to use the legacy policy but allow you to enable the enhanced policy if you want to test something
* Don't set `GTF_CALL_M_DOES_NOT_RETURN` if the legacy policy is in use
* Don't set `fgRemoveRestOfBlock` if the call type is not void
* Added comment to the benchmark to explain its purpose
* Rebased

I did a diff between the previous version (which handled non-void calls) and this updated version and there were 5 bytes of diff from 2 callsites. In both cases the methods throw "not supported" like exceptions and the lack of "no return" information affects only the exceptional path. It doesn't seem worthwhile to handle such rare cases.

AndyAyersMS commented on src/jit/flowgraph.cpp:
```diff
@@ -5768,14 +5768,51 @@ void          Compiler::fgFindBasicBlocks()
+                impInlineInfo->iciCall->gtCallMoreFlags |= GTF_CALL_M_DOES_NOT_RETURN;
+            }
+
+            compInlineResult->NoteBool(InlineObservation::CALLEE_DOES_NOT_RETURN, !hasReturnBlocks);
```
I realize I'm being picky, but 'd prefer that observations be made unconditionally. So can you restructure this bit as follows?

```C++
if (!hasReturnBlocks)
{
    if(!compInlineResult->UsesLegacyPolicy())
    {
        // ... comment ...
        impInlineInfo->iciCall->gtCallMoreFlags |= GTF_CALL_M_DOES_NOT_RETURN;
    }
    compInlineResult->NoteBool(InlineObservation::CALLEE_DOES_NOT_RETURN);
    if (compInlineResult->IsFailure())
    {
        return;
    }
}
```

mikedn replied:
More like the following I'd say:
```CPP
if (!hasReturnBlocks && !compInlineResult->UsesLegacyPolicy())
{
    // ...
    impInlineInfo->iciCall->gtCallMoreFlags |= GTF_CALL_M_DOES_NOT_RETURN;
}
compInlineResult->NoteBool(InlineObservation::CALLEE_DOES_NOT_RETURN, !hasReturnBlocks);
if (compInlineResult->IsFailure())
{
    return;
}
```
`NoteBool` has to be called no matter what value `hasReturnBlocks` has, otherwise an assert fires in the enhanced policy.

AndyAyersMS replied:
Right, I should have remembered.

Your version looks good.

AndyAyersMS commented on src/jit/jitconfigvalues.h:
```diff
@@ -203,6 +203,12 @@ CONFIG_STRING(JitNoInlineRange, W("JitNoInlineRange"))
+#else
+CONFIG_INTEGER(JitInlinePolicyLegacy, W("JitInlinePolicyLegacy"), 1)
+#endif
+
```
Please enable the new policy by default even for non-CoreCLR. We will opt out later if we determine we don't want this new behavior for desktop.

AndyAyersMS commented:
X86 failure is `GC\Scenarios\DoublinkList\dlstack` which has been failing sporadically recently in various PRs.

I'm less sure about the arm failures, but we didn't see them in any of the earlier runs of this and things haven't changed that much.

So am going to retry these... will drill in more if they repro

@dotnet-bot retest Linux ARM Emulator Cross Debug Build
@dotnet-bot retest Linux ARM Emulator Cross Release Build
@dotnet-bot retest Windows_NT x86 ryujit Checked Build and Test

AndyAyersMS commented:
None of the failures repro. Changes look good.

AndyAyersMS commented:
@dotnet/jit-contrib FYI this is going to alter default inliner behavior. Will be merging it shortly.

benaadams commented:
Woop! :+1:

benaadams commented:
Woop! :+1: 

Is the jit used to create the native image? (e.g. System.Private.CoreLib.ni.dll)

AndyAyersMS commented:
Yes, it is.

mikedn commented:
@AndyAyersMS Thanks!

@benaadams I'm not sure what extra optimizations you have in mind. I suppose in some cases preventing unnecessary inlining of such methods may allow other method to be inlined. But I don't remember seeing anything significant in the diffs, this is primarily a code size optimization.

benaadams commented:
@mikedn changing repeated direct throws to functions. 

Unfortunately I can't seem to get jit-diff to work to say if this would be a positive change (thought the il is different); but for example:

https://github.com/benaadams/coreclr/commit/ab5e535b40e43f70e2127d7fef7061ed5f71ec5b

AndyAyersMS commented:
There may be some subtleties in getting jit-diff to measure size changes caused by changes in the core library sources (holding the jit constant), since that's not the default workflow.

I think it should be doable in steps -- collect the base and diff results separately making sure to get the right corlib versions for each,  and then analyze them using jit-analyze. I can give it a try later.

benaadams commented:
@AndyAyersMS ah, I went too deep :) I'll try something similar in corefx

mikedn commented:
> Unfortunately I can't seem to get jit-diff to work to say if this would be a positive change

Yes, those kind of changes result in native code size reductions. corelib already uses such throw helpers in a few places, that's why the jit-diff summary I posted previously shows a ~100kb reduction in code size.

For generic types/methods it's possible to not see any reductions in the native image size if the input assembly doesn't instantiate those generic types/methods anywhere. Still, the code generated at runtime should be smaller.

AndyAyersMS commented:
@benaadams here's what I get from jit-diffs based on benaadams@ab5e535. I belive jit-diff is actually folding overloads together and reporting results under the name of the first overload it sees -- which is why the improvment numbers for the methods are so high -- but I'll need to investigate further to be sure.
```
Total bytes of diff: -39686 (-1.23 % of base)
    diff is an improvement.

Total byte diff includes -9556 bytes from reconciling methods
        Base had  135 unique methods,    10076 unique bytes
        Diff had    8 unique methods,      520 unique bytes

Top file improvements by size (bytes):
      -39686 : System.Private.CoreLib.dasm (-1.23 % of base)

1 total files with size differences.

Top method regessions by size (bytes):
         261 : System.Private.CoreLib.dasm - System.Reflection.CustomAttributeData:Init(ref):this
          75 : System.Private.CoreLib.dasm - System.Array:ThrowArgumenCountOutOfRange()
          75 : System.Private.CoreLib.dasm - System.Array:ThrowArgumenStartIndexOutOfRange()
          75 : System.Private.CoreLib.dasm - System.Array:ThrowArgumentElementTypeMustBeType()
          70 : System.Private.CoreLib.dasm - System.Reflection.CustomAttributeData:GetCustomAttributesInternal(ref):ref

Top method improvements by size (bytes):
       -2801 : System.Private.CoreLib.dasm - System.Array:LastIndexOf(ref,struct,int,int):int
       -2328 : System.Private.CoreLib.dasm - System.Array:Sort(ref,int,int,ref)
       -1996 : System.Private.CoreLib.dasm - System.Array:IndexOf(ref,struct,int,int):int
       -1979 : System.Private.CoreLib.dasm - System.Internal:SZArrayHelper(ref)
       -1800 : System.Private.CoreLib.dasm - System.SZArrayHelper:GetEnumerator():ref:this

253 total methods with size differences.
```

russellhadley commented:
@AndyAyersMS, Yes the tool will fold all overloads together.  This is some legacy behavior from an internal tool which was kept for convenience.  The data that there were multiple instances is captured in the data structures in the tool so we could add some kind of notation to mark multiple instance entries.

benaadams commented:
@AndyAyersMS with https://github.com/dotnet/coreclr/pull/6634 which has a couple more I'm seeing a 68,608 byte reduction of filesize

AndyAyersMS commented:
@russellhadley I don't find folding to be a particularly useful default -- seems like instead we should report actual per-method data and provide aggregate views (by method, by class, by assembly, etc) to summarize.

AndyAyersMS commented:
@russellhadley Hmm, I suppose we're stuck with some degree of folding since the jit dump has simplified types like ref, struct, etc. I'll update the reporting to at least show the number of folded instances so it's obvious there is more than one method being aggregated.

briansull commented:
Ideally we would like to be able to quickly find the top 10 methods that improved and the top 10 methods that regressed. It seems like aggregating multiple copies of a method together makes that task difficult.  Is `LastIndexOf` really the method with the best improvement if there were 200 copies of that method aggregated, probably not.

AndyAyersMS commented:
Here's the result with a bit of tweaking to `jit-analyze`. May not be 100% accurate if you add or remove overloads, but better than nothing.

Note I am still looking at CHK builds of corlib since we still need to sort out how to get the tooling to do the right jit swaps to measure the REL corlib (see dotnet/jitutils#14).
```
Top method improvements by size (bytes):
       -2801 : System.Private.CoreLib.dasm - System.Array:LastIndexOf(ref,struct,int,int):int (9 methods)
       -2328 : System.Private.CoreLib.dasm - System.Array:Sort(ref,int,int,ref) (25 methods)
       -1996 : System.Private.CoreLib.dasm - System.Array:IndexOf(ref,struct,int,int):int (11 methods)
       -1979 : System.Private.CoreLib.dasm - System.Internal:SZArrayHelper(ref) (19 methods)
       -1800 : System.Private.CoreLib.dasm - System.SZArrayHelper:GetEnumerator():ref:this (19 methods)
```

AndyAyersMS commented:
PR for tool update is dotnet/jitutils#22

mikedn commented:
> I belive jit-diff is actually folding overloads together and reporting results under the name of the first overload it sees -- which is why the improvment numbers for the methods are so high --

Note that this also happens for generic instantiations. That's how `IfNullAndNullsAreIllegalThenThrow` ended up being 3345 bytes smaller even if the method itself is quite small.
